File: MVC Lab/Guitar Example/C#/GuitarExample/GuitarExample/Controllers/GuitarsController.cs
using System.Collections.Generic;
using System.Linq;
using System.Web.Mvc;
using GuitarExample.Models;

namespace GuitarExample.Controllers
{
    public class GuitarsController : Controller
    {
        protected IList<Guitar> Guitars = new List<Guitar>();

        public GuitarsController()
        {
            if (System.IO.File.Exists(@"c:\temp\Guitars.txt"))
            {
                var fileContent = System.IO.File.ReadAllText(@"c:\temp\Guitars.txt");
                Guitars = fileContent.Split(',').Where(x => !string.IsNullOrEmpty(x.Trim())).Select(x => new Guitar {Name = x}).ToList();
            }
        }

        public ActionResult Index()
        {
            return View(Guitars);
        }

        [HttpGet]
        public ActionResult Create()
        {
            return View();
        }

        [HttpPost]
        public ActionResult Create(Guitar guitar)
        {
            var isNameValue = !string.IsNullOrEmpty(guitar.Name);
            if (!isNameValue && !ModelState.IsValid || isNameValue && !ModelState.IsValid || !isNameValue && ModelState.IsValid)
            {
                return View(guitar);
            }
            else
            {
                var content = Guitars.Aggregate("", (acc, x) => acc + x.Name + ",");
                System.IO.File.WriteAllText(@"c:\temp\Guitars.txt", content + guitar.Name);
                return RedirectToAction("Index");
            }
        }
    }
}

File: MVC Lab/Guitar Example/F#/GuitarExample/GuitarExample/GuitarExampleWebApp/GuitarsController.fs
namespace FsWeb.Controllers

open System
open System.Web.Mvc
open FsWeb.Models

[<HandleError>]
type GuitarsController() =    
    inherit Controller()    

    let guitars = match System.IO.File.Exists @"c:\temp\Guitars.txt" with
                  | true -> 
                      System.IO.File.ReadAllText(@"c:\temp\Guitars.txt").Split(',') 
                      |> Array.map (fun x -> Guitar(Name = x))
                  | _ -> [||]

    member this.Index () = guitars |> this.View 

    [<HttpGet>] 
    member this.Create () = this.View() 
    
    [<HttpPost>] 
    member this.Create (guitar : Guitar) : ActionResult =
        let isNameValid = not (String.IsNullOrEmpty(guitar.Name))

        match base.ModelState.IsValid, isNameValid with
        | false, false | true, false | false, true -> 
            upcast this.View guitar
        | _ -> 
            let result = guitars |> Array.fold(fun acc x -> acc + x.Name + ",") ""
            System.IO.File.WriteAllText(@"c:\temp\Guitars.txt", result + guitar.Name)
            upcast base.RedirectToAction("Index") 
